### ISSUE TITLE
do not require type and allow giving optional “schema” object

### ISSUE THREAD

**Issue by kanitw** [vega/vega-lite#512]
- [ ] list primitive type (string, integer, number, date time, etc.)
- [ ] The schema should list natural order of an ordinal field.   (e.g., “small”, “medium”, “large”)
- [ ] it would be useful to optionally have default aggregation function, numberFormat, color range as well.
- [ ] description

**Comment by kanitw** [comment]
this needs to be revisited as we are considering to remove stats #648

**Comment by domoritz** [comment]
Let's close this as we have a better idea what we want in #648

**Comment by kanitw** [comment]
I disagree, this issue is about making type optional, which is not necessary a part of eliminating stats.

**Comment by rgrp** [comment]
@domoritz would it be possible to be using JSON Table Schema stuff here -- what's missing from [JSON Table Schema](http://dataprotocols.org/json-table-schema/)? I guess my bigger question is whether you could replace your data source object with a [Data Package resource object](http://dataprotocols.org/data-packages/#resource-information)?

I ask because this is naturally what I would want to do in: https://github.com/dataprotocols/dataprotocols/issues/77 -- basically I want data specified by Data Package then connected with visualization specified by vega-lite. At the moment, obviously you end up defining your own data source mini-language and I wonder if we could converge there replacing vega-lite's data source with a subpart of Data Package spec and Data Package using vega-lite for view specs.

**Comment by domoritz** [comment]
@rgrp I was out of the country for a few weeks. Please excuse the late reply.

The data source object uses a similar API to vega (https://github.com/vega/vega/wiki/Data). Because of that, I don't see the benefit of using JSON Table Schema (or actually only a small subset, we would at most rename one or two properties). What exactly do you think we would need to change? 

I think that data packages could use a vega-lite spec without a data object. When you want to visualize a data package, the application would then generate a simple data source object either with embedded data (parsed using the data package library) or a simple url. From what I read, I don't think we will ever be able to support the full data package resource objects.

**Comment by kanitw** [comment]
- [ ] The schema should list natural order of an ordinal field.   (e.g., “small”, “medium”, “large”)

**Comment by kanitw** [comment]
- [ ] it would be useful to optionally have default aggregation function, numberFormat, color range as well.

**Comment by rgrp** [comment]
@domoritz let me try and set out the logic again a bit more clearly:

* vega / vega-lite is a system for specifying **visualizations**
* to create visualizations you normally have (at least) the following 3 components
  * general-metadata - e.g. title of graph, credits …
  * data
  * graph description / specification
* data itself often consists of 2 things really in most implementations:
  * "raw-data": that is data exactly in the form needed by the visualization system. This is often a very well defined spec e.g. an array of series … 
  * data-import + transform: this is very common. For the convenience of users one supports importing - and transforming - data from external or other sources e.g. a CSV, or even a JSON array (which is not yet in the "raw-data" form)
* As is usual you've ended up writing your own mini-data spec (and import/transform spec) into vega-lite even though vega-lite is really about the visualization site - the "graph description / specification"
* This means you take up time working out your data work and fixing your import and transform sytem - as evidenced by this very issue or the discussion about units in #817 or several other issues.
* **What I am suggesting**:
  * focus vega-lite purely on the "graph description / specification"
  * define a simple raw-data spec sufficient for vega-lite needs - this is the crucial "interface" between vega-lite and any external data sources. If you can get your data into this "raw-data" form then vega-lite can use it
  * "reuse" for your more general data spec (the import / transform / import) to JSON Table Schema and the resource definition in Data Package. 
   * And we together focus on getting data from JSON Table Schema / Resource into the "raw-data" form that vega-lite needs

In summary, I'm suggesting a "separation of concerns". Rather than vega-lite inventing its own new data definition system including referencing external sources, having units, setting data types etc etc you reuse the existing JSON Table Schema and related specifications.

**Comment by domoritz** [comment]
Hi @rgrp. Vega-Lite is not a charting library. It is designed to be a specification language for _visual analysis_. 

Since it is designed for analysis, transformation like filtering, aggregating and binning becomes a critical piece of the language. Tableau is one of our inspirations and an example of a visualization tool that is designed for analysis and thus supports transformations as its first class features. We want to make it as easy as possible for people to see their data rather than having to combine multiple libraries to first download the data and then aggregate/visualize it. 

> Rather than vega-lite inventing its own new data definition system including referencing external sources, having units, setting data types etc etc you reuse the existing JSON Table Schema and related specifications.

We are not inventing our new data definition system. Vega-Lite compiles to Vega, which already has defined how it loads data. Since Vega has been release a few years ago, replacing Vega's data system isn't a logical option. 

The data types we have in Vega-Lite are not types like int, float, string, ... but instead [Levels of Measurement](https://en.wikipedia.org/wiki/Level_of_measurement) with extension to support temporal data. That said, for this issue, it is very likely that our schema support would be very close to JSON Table Schema.  However, as I said earlier, some other properties in the schema doesn't match our design.  For example, for us, `format` property should match d3's format which is the underlying representation.  Trying to provide 1:1 mapping from JSON table schema to d3 format isn't on our plan.

Another useful part is that Vega actually supports loading data inline so that would be useful for a use case where you already have your data transformed.  

> This means you take up time working out your data work and fixing your import and transform sytem - as evidenced by this very issue or the discussion about units in #817 or several other issues.

It is correct that we don't want to spent much effort on supporting many data sources and formats. However, some properties of the data such as units affect how we display the data.

Concerning your suggestions, we could change them a little bit to make data protocols and vega lite compatible. 

> focus vega-lite purely on the "graph description / specification"

As explained earlier, we won't be able to do that.

> define a simple raw-data spec sufficient for vega-lite needs - this is the crucial "interface" between vega-lite and any external data sources. If you can get your data into this "raw-data" form then vega-lite can use it

Vega-Lite compiles to Vega. The [runtime](https://github.com/vega/vega/wiki/Runtime#basic-use) already supports injecting data at runtime. 

What I'm suggesting is that Vega-Lite and Vega stays as they are. They support the raw data embedded in the spec (`values`) and simple data loading from a URL (JSON and CSV). A data package could be as you suggested with added Vega-Lite.

```
{
  "name": "my-data-package",
  "resources": [
    {
      "name": "my-csv.csv",
      "schema": [
        "fields": [ { "name": "column1", "type": number"}, ....]
    }
  ]
  "views": [
    {
      "title": "...",
      "data": { "resource": ..., "query": .... },
      "spec": {
        "mark": "point",
        "encoding": {
          "x": {"field": "Horsepower", "type": "quantitative"}
        }
      }
    }
  ]
}
```

Note that I left out the at a from the Vega-Lite spec. Then the library for data packages could then run the following (pseudo) code.

```js
loader = require("data-package-loader");
vl = require("vega-lite");
vg = require("vega");

package.views.forEach((view)=> {
  data = loader.downloadAndParse(view.data);
  vgSpec = vl.compile(vlSpec).spec;
  vg.parse.spec(vgSpec, function(chart) {
    chart({el:"#vis", data: data}).update();
  });
});
```

Would this be the separation of concerns that you're suggesting? The data packages library could take care of the parsing and loading part and then pass it to Vega. We wouldn't use the JSON table schema for Vega/Vega-Lite but instead wrap the visualization specification part in a data package and the library that loads the data package can inject the data into the visualization.

**Comment by rgrp** [comment]
> Vega-Lite is not a charting library. It is designed to be a specification language for visual analysis.

Yes ;-) Really appreciate the detailed response.

> Since it is designed for analysis, transformation like filtering, aggregating and binning becomes a critical piece of the language. 

Right, but the question is can you reuse stuff from elsewhere rather than rolling your own for everything ;-) -- just like in Data Packages we can reuse vega-lite you could reuse some of the data import, data definition and data transform from elsewhere incorporating it in.

> However, some properties of the data such as units affect how we display the data.

I know and there is units work going on in data protocols and data packages - is it possible to join forces to get a common solution that works for both of us?

> For example, for us, format property should match d3's format which is the underlying representation. Trying to provide 1:1 mapping from JSON table schema to d3 format isn't on our plan.

could you piont to the d3 format field so we could get a sense of what is going on there.

> Another useful part is that Vega actually supports loading data inline so that would be useful for a use case where you already have your data transformed.

Data Package resources support inline data btw - not sure if that was clear.

> > focus vega-lite purely on the "graph description / specification"
> 
> As explained earlier, we won't be able to do that.

My point is not that you don't have this stuff in vega-lite but that you reuse other specs for the parts where you need it. Does that make more sense. The reason here is self-interest as you need to work out less and can focus on those parts you really need. It also makes it more likely other reuse code increasing adoption etc.

## Example

Your example about "separation of concerns" and reuse is pretty much spot-on. For me the main thing is to:

* understand what part of the vega-lite spec the `spec` attribute should point to
* what exact form we should get data into for use e.g. how should we transform resource data so it is usable i.e. in your example the algorithm for the `loader.downloadAndParse(view.data);`

**Comment by domoritz** [comment]
> could you piont to the d3 format field so we could get a sense of what is going on there.

Note that vega and vega-lite and heavily relying on [datalib](http://vega.github.io/datalib/). You might want to check it out to see what parsing we already support. So Vega passes parsing down to datalib and datalib uses D3 format for example for date parsing. Documentation starts at https://github.com/vega/vega/wiki/Data#-json and you can find format patters in https://github.com/vega/datalib/wiki/Import#dl_read.

> understand what part of the vega-lite spec the spec attribute should point to

I'd say the complete vega-lite spec without the `data` property. 

> what exact form we should get data into for use e.g. how should we transform resource data so it is usable i.e. in your example the algorithm for the loader.downloadAndParse(view.data);

Vega expects js objects that have an object for each row.

```js
[{foo: 1, bar: 3}, {foo: 4, bar: 2}, {foo: 4, bar: 6}, {foo: 2, bar: 7}]
```

I think that should be quite easy to support.

**Comment by rgrp** [comment]
## Data Libraries and Specifications

> Note that vega and vega-lite and heavily relying on datalib. You might want to check it out to see what parsing we already support. So Vega passes parsing down to datalib and datalib uses D3 format for example for date parsing. Documentation starts at https://github.com/vega/vega/wiki/Data#-json and you can find format patters in https://github.com/vega/datalib/wiki/Import#dl_read.

@domoritz I think one point might be datalib supporting JTS or even the "resource" objects you have in Tabular Data Packages (which has both schema and dialect descriptor format for CSV). This is just a suggestion. I also wonder if there might be some mileage in working out any overlap between datalib and recline - perhaps that could become something common.

The logic for all of this is to avoid people having to all maintain there own data libraries.

The process:

```CSV / some other format -> array of json objects```

Is probably quite a standard one and it is one that e.g. recline solves, datalib solves, https://github.com/datapackages/tabulator-py solves etc. Whilst the simple part of this is very simple - as you add functionality (type sniffing, type casting, rudimentary validation, data transformation) it gets more complex - and there is always some benefit in sharing some stuff (be an actual library or just some standards like JSON Table Schema).

To repeat my comment a couple of comments back:

> > Since it is designed for analysis, transformation like filtering, aggregating and binning becomes a critical piece of the language.
>
> Right, but the question is can you reuse stuff from elsewhere rather than rolling your own for everything ;-) -- just like in Data Packages we can reuse vega-lite you could reuse some of the data import, data definition and data transform from elsewhere incorporating it in.

And 

> > However, some properties of the data such as units affect how we display the data.
>
> I know and there is units work going on in data protocols and data packages - is it possible to join forces to get a common solution that works for both of us?

### Using vega-lite in data packages

I'm going to take a look at the "Tabular Data Package Resource --> Vega-lite data structure" pretty soon and start trialling use of vega. Will keep you posted.

/cc @pwalsh

**Comment by domoritz** [comment]
> I'm going to take a look at the "Tabular Data Package Resource --> Vega-lite data structure" pretty soon and start trialling use of vega. Will keep you posted.

Excellent. Let us know if you have questions.

> The logic for both of these is save you (and all of us) each maintaining our own independent libraries doing roughly the same thing e.g.

Totally. Datalib actually uses D3 to parse CSV (https://github.com/vega/datalib/blob/master/src/import/formats/dsv.js#L2). We are relying on a number of d3 libraries for basic things.

**Comment by kanitw** [comment]
This issue has been around for 3 years and we probably won't get to do it as there are many other more important issues. So I'm closing this.
